Use Regular Expressions to validate Assessment Identifiers

JCantu248 opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

Gophish-export is modified to validate assessment_id input prior to processing. This corrects some errors in processing that can come from invalid or malformed input. 

## 💭 Motivation and context ##

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

This change is in relation to https://github.com/cisagov/gophish-tools/issues/11 .  A partial assessment_id like "RV" could be passed to gophish-export, resulting in it matching all campaigns beginning in RV (which is all campaigns for all assessments). A regex validator was added to validate that the assessment_id matches the known format of assessments. 

Closes #11 

## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->

<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

Gophish-export was run with ``assessment_id = "RV"`` and confirmed that an error message was shown to the user, and the process would exit. 

Pytest cases were added to validate that the Regex matched a valid assessment ID format, and that it would not match an invalid value. 

## 📷 Screenshots (if appropriate) ##

<!-- Remove this section and header if not needed -->

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

JCantu248 committed:
Adding regular expression to validate that assessment_id.

Gophish-export allows for an invalid assessment_ids that can lead to unexpected behavior. Regex was added to ensure assessment_id matches the format RV1234.

JCantu248 committed:
Bump project version from 0.0.7 to 0.0.8

JCantu248 committed:
Bump project version from 0.0.8 to 0.0.8-rc.1

JCantu248 committed:
Bump project version from 0.0.8-rc.1 to 0.0.9

JCantu248 committed:
Bump project version from 0.0.9 to 0.0.9-rc.1

JCantu248 committed:
Bump project version from 0.0.9-rc.1 to 0.0.9-rc.2

JCantu248 committed:
Bump project version from 0.0.9-rc.2 to 0.0.9-rc.3

JCantu248 committed:
Debugging issues with new regex for assessment_id validation

JCantu248 committed:
Bump project version from 0.0.9-rc.4 to 0.0.9-rc.5

JCantu248 committed:
Spun regex validation to its own function.

JCantu248 committed:
Bump project version from 0.0.9-rc.6 to 0.0.9

JCantu248 committed:
Moved validate_assessment_id to util/validate.py and added test cases

JCantu248 committed:
Bump project version from 0.0.9-rc.7 to 0.0.9

JCantu248 committed:
Accidentally incremented the version number from 0.0.7 to 0.0.9 somewhere during the course of development. Correcting it back to 0.0.8 for release.

jsf9k commented on src/util/validate.py:
```diff
@@ -15,6 +15,21 @@
+    Returns:
+        match: the result of a regular expression match.
+    """
+    pattern = re.compile("^RV([0-9]){4}")
+    match = pattern.match(assessment_id)
```
You are re-compiling the regex every time you call the expression anyway, so you may as well use the one-liner syntax.  Also, the parentheses inside the regex are unnecessary, and you can replace `[0-9]` with the pithier `\d`:
```suggestion
    pattern = re.match(r"^RV\d{4}", assessment_id)
```

JCantu248 replied:
See 0c1b4765449b645fc9b7b00779039ddee6410a92 for changes

JCantu248 replied:
Hey @jsf9k, is this one resolved?

jsf9k replied:
Yep, sorry I overlooked it.

jsf9k reviewed (commented):
I made a few suggestions to simplify your regex pattern match.

JCantu248 committed:
Change regex validation to a one liner, simplify regex expression.

dav3r commented on src/util/validate.py:
```diff
@@ -15,6 +15,20 @@
+        assessment_id (string): Assessment identifier to validate.
+
+    Returns:
+        match: the result of a regular expression match.
```
Use consistent capitalization:
```suggestion
        match: The result of a regular expression match.
```

JCantu248 replied:
See c6d29c5 for revisions

dav3r commented on src/util/validate.py:
```diff
@@ -15,6 +15,20 @@
+    Returns:
+        match: the result of a regular expression match.
+    """
+    match = re.match(r"^RV\d{4}", assessment_id)
+
+    return match
```
Since this is now a one-liner, we might as well eliminate the `match` variable altogether:
```suggestion
    return re.match(r"^RV\d{4}", assessment_id)
```

JCantu248 replied:
See c6d29c5 for revisions

dav3r commented on src/util/validate.py:
```diff
@@ -15,6 +15,20 @@
+    Returns:
+        match: the result of a regular expression match.
+    """
+    match = re.match(r"^RV\d{4}", assessment_id)
```
What happens when assessments IDs are longer than 4 digits?

JCantu248 replied:
Ben and I talked about this. We figure we will be using the 4 digit RV numbers for a few years, so he thought it was okay. If we end up changing the format, we would have to revise this.

dav3r replied:
I just realized that your regex will still match assessment IDs like `RV1234ABC` and `RV123456789`, so it should probably include a `$` like so:
```
match = re.match(r"^RV\d{4}$", assessment_id)
```

Even better (in my opinion) would be to match 4 or 5 digit assessment IDs to future-proof this:
```
match = re.match(r"^RV\d{4,5}$", assessment_id)
```

JCantu248 replied:
The regex was modified and included in the previous change to make the function a one-liner. 

See c6d29c5 for revisions

dav3r commented on src/tools/gophish_export.py:
```diff
@@ -400,6 +401,13 @@ def main() -> None:
 
+    if not validate_assessment_id(args["ASSESSMENT_ID"]):
+        logging.critical(
+            '"%s" is an invalid assessment_id format. Example: RV1234',
+            args["ASSESSMENT_ID"],
```
This error message could be a bit more explanatory about what the valid format is, rather than just giving the example `RV1234`.

JCantu248 replied:
Error message was modified to explain the assessment_id format.

See c6d29c5 for revisions

dav3r reviewed (changes requested):
Please consider my comments and questions.

JCantu248 committed:
Expanded the regex validatator to accept either 4 or 5 digit assessment_ids.

Other small changes to error prompt when receiving malformed input, and revising validate_assessment_id() to simplify it down to a one line function.

dav3r commented on src/tools/gophish_export.py:
```diff
@@ -400,6 +401,14 @@ def main() -> None:
+    if not validate_assessment_id(args["ASSESSMENT_ID"]):
+        logging.critical(
+            '"%s" is an invalid assessment_id format. Assessment Identifiers begin with RV and followed by a 4 or 5 '
+            "digit numerical sequence. Example: RV1234",
```
Small capitalization/grammar cleanup:
```suggestion
            '"%s" is an invalid assessment_id format. Assessment identifiers begin with RV and are followed by '
            " a 4 or 5 digit numerical sequence. Examples: RV1234, RV12345",
```

JCantu248 replied:
See here for changes applied 

9c83ccb

JCantu248 committed:
Revised error message according to feedback

dav3r reviewed (approved):
Thanks for making the changes I requested - looks good! 🏎️

JCantu248 committed:
Merging changes in develop branch into this feature branch prior to merger.

Merge branch 'develop' of github.com:cisagov/gophish-tools into Issue_11_AssessmentID_match_fix
